src/drivers/magnetometer/qmc5883/qmc5883.cpp

```diff
@@ -0,0 +1,1179 @@
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file qmc5883.cpp
+ *
+ * Driver for the QMC5883 magnetometer connected via I2C or SPI.
+ */
+
+#include <px4_config.h>
+#include <px4_defines.h>
+
+#include <drivers/device/i2c.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <semaphore.h>
+#include <string.h>
+#include <fcntl.h>
+#include <poll.h>
+#include <errno.h>
+#include <stdio.h>
+#include <math.h>
+#include <unistd.h>
+
+#include <nuttx/arch.h>
+#include <nuttx/wqueue.h>
+#include <nuttx/clock.h>
+
+#include <board_config.h>
+
+#include <perf/perf_counter.h>
+#include <systemlib/err.h>
+
+#include <drivers/drv_mag.h>
+#include <drivers/drv_hrt.h>
+#include <drivers/device/ringbuffer.h>
+#include <drivers/drv_device.h>
+
+#include <uORB/uORB.h>
+
+#include <float.h>
+#include <getopt.h>
+#include <lib/conversion/rotation.h>
+
+#include "qmc5883.h"
+
+/*
+ * QMC5883 internal constants and data structures.
+ */
+
+/* Max measurement rate is 200Hz */
+#define QMC5883_CONVERSION_INTERVAL	(1000000 / 150)	/* microseconds */
+#define QMC5883_MAX_COUNT 			32767
+
+#define QMC5883_ADDR_DATA_OUT_X_LSB		0x00
+#define QMC5883_ADDR_DATA_OUT_X_MSB		0x01
+#define QMC5883_ADDR_DATA_OUT_Y_LSB		0x02
+#define QMC5883_ADDR_DATA_OUT_Y_MSB		0x03
+#define QMC5883_ADDR_DATA_OUT_Z_LSB		0x04
+#define QMC5883_ADDR_DATA_OUT_Z_MSB		0x05
+#define QMC5883_ADDR_STATUS			0x06
+#define QMC5883_ADDR_TEMP_OUT_LSB 		0x07
+#define QMC5883_ADDR_TEMP_OUT_MSB 		0x08
+#define QMC5883_ADDR_CONTROL_1 			0x09
+#define QMC5883_ADDR_CONTROL_2 			0x0A
+#define QMC5883_ADDR_SET_RESET 			0x0B
+
+
+#define QMC5883_STATUS_REG_DRDY 		(1 << 0)  /* Data Ready: "0": no new data, "1": new data is ready */
+#define QMC5883_STATUS_REG_OVL 			(1 << 1)  /* Overflow Flag: "0": normal, "1": data overflow */
+#define QMC5883_STATUS_REG_DOR			(1 << 2)  /* Data Skip: "0": normal, "1": data skipped for reading */
+
+/* Control Register 1 */
+#define QMC5883_MODE_REG_STANDBY 		(0 << 0)
+#define QMC5883_MODE_REG_CONTINOUS_MODE         (1 << 0)
+#define QMC5883_OUTPUT_DATA_RATE_10 		(0 << 2) /* Hz */
+#define QMC5883_OUTPUT_DATA_RATE_50 		(1 << 2)
+#define QMC5883_OUTPUT_DATA_RATE_100 		(2 << 2)
+#define QMC5883_OUTPUT_DATA_RATE_200 		(3 << 2)
+#define QMC5883_OUTPUT_RANGE_2G 		(0 << 4)  /* +/- 2 gauss */
+#define QMC5883_OUTPUT_RANGE_8G 		(1 << 4)  /* +/- 8 gauss */
+#define QMC5883_OVERSAMPLE_512 			(0 << 6)  /* controls digital filter bw - larger OSR -> smaller bw */
+#define QMC5883_OVERSAMPLE_256 			(1 << 6)
+#define QMC5883_OVERSAMPLE_128 			(2 << 6)
+#define QMC5883_OVERSAMPLE_64 			(3 << 6)
+
+/* Control Register 2 */
+#define QMC5883_INT_ENB 			(1 << 0)
+#define QMC5883_ROL_PNT 			(1 << 6)
+#define QMC5883_SOFT_RESET 			(1 << 7)
+
+/* Set Register */
+#define QMC5883_SET_DEFAULT 			(1 << 0)
+
+#define QMC5883_TEMP_OFFSET 50 /* deg celsius */
```

Comment:
Temp offset is not calibrated on sensor at the factory so this is a pretty rough estimate to get ballpark sensor temp.  All temp calibration is happening on chip by default.